core/trie2/triedb/pathdb/layertree_test.go

```diff
@@ -0,0 +1,495 @@
+		classPaths:           make(map[trieutils.Path]trienode.TrieNode),
+		contractPaths:        make(map[trieutils.Path]trienode.TrieNode),
+		contractStoragePaths: make(map[felt.Felt]map[trieutils.Path]trienode.TrieNode),
+	}
+}
+
+func (t *layerTracker) trackLayer(root, parent felt.Felt) {
+	if root.Equal(&parent) {
+		return
+	}
+	t.childToParent[root] = parent
+}
+
+func (t *layerTracker) trackNodes(
+	root felt.Felt,
+	parent felt.Felt,
+	classNodes,
+	contractNodes map[trieutils.Path]trienode.TrieNode,
+	storageNodes map[felt.Felt]map[trieutils.Path]trienode.TrieNode,
+) {
+	t.trackClassNodes(root, classNodes)
+	t.trackContractNodes(root, contractNodes)
+	t.trackContractStorageNodes(root, storageNodes)
+	t.trackLayer(root, parent)
+}
+
+func (t *layerTracker) trackClassNodes(root felt.Felt, nodes map[trieutils.Path]trienode.TrieNode) {
+	for path, node := range nodes {
+		if t.classNodes[root] == nil {
+			t.classNodes[root] = make(map[trieutils.Path]trienode.TrieNode)
+		}
+		t.classNodes[root][path] = node
+		t.classPaths[path] = node
+	}
+}
+
+func (t *layerTracker) trackContractNodes(root felt.Felt, nodes map[trieutils.Path]trienode.TrieNode) {
+	for path, node := range nodes {
+		if t.contractNodes[root] == nil {
+			t.contractNodes[root] = make(map[trieutils.Path]trienode.TrieNode)
+		}
+		t.contractNodes[root][path] = node
+		t.contractPaths[path] = node
+	}
+}
+
+func (t *layerTracker) trackContractStorageNodes(root felt.Felt, nodes map[felt.Felt]map[trieutils.Path]trienode.TrieNode) {
+	for owner, ownerNodes := range nodes {
+		if t.contractStorageNodes[root] == nil {
+			t.contractStorageNodes[root] = make(map[felt.Felt]map[trieutils.Path]trienode.TrieNode)
+		}
+		if t.contractStorageNodes[root][owner] == nil {
+			t.contractStorageNodes[root][owner] = make(map[trieutils.Path]trienode.TrieNode)
+		}
+		for path, node := range ownerNodes {
+			t.contractStorageNodes[root][owner][path] = node
+			if t.contractStoragePaths[owner] == nil {
+				t.contractStoragePaths[owner] = make(map[trieutils.Path]trienode.TrieNode)
+			}
+			t.contractStoragePaths[owner][path] = node
+		}
+	}
+}
+
+// resolveNode finds a node by traversing the layer hierarchy from the given root
+func (t *layerTracker) resolveNode(root, owner felt.Felt, path trieutils.Path, isClass bool) ([]byte, error) {
+	currentRoot := root
+	for {
+		if blob, found := t.findNodeInLayer(currentRoot, owner, path, isClass); found {
+			return blob, nil
+		}
+
+		// Try parent layer if available
+		parent, hasParent := t.childToParent[currentRoot]
+		if !hasParent {
+			return nil, fmt.Errorf("node not found in layer hierarchy: root=%v, owner=%v, path=%v",
+				root.String(), owner.String(), path.String())
+		}
+		currentRoot = parent
+	}
+}
+
+// findNodeInLayer checks if a node exists in a specific layer (without parent traversal)
+func (t *layerTracker) findNodeInLayer(root, owner felt.Felt, path trieutils.Path, isClass bool) ([]byte, bool) {
```

Comment:
`root`, `owner` and `path` by reference please

Reply:
done